pkg/infrastructure/gcp/clusterapi/dns.go

```diff
@@ -0,0 +1,178 @@
+package clusterapi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/api/dns/v1"
+
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	gcpic "github.com/openshift/installer/pkg/asset/installconfig/gcp"
+	"github.com/openshift/installer/pkg/types"
+)
+
+var (
+	errNotFound = errors.New("not found")
+)
+
+func getDNSZoneName(ctx context.Context, ic *installconfig.InstallConfig, isPublic bool) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+
+	client, err := gcpic.NewClient(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to create new client: %w", err)
+	}
+
+	cctx, ccancel := context.WithTimeout(ctx, time.Minute*1)
+	defer ccancel()
+
+	domain := ic.Config.ClusterDomain()
+	if isPublic {
+		domain = ic.Config.BaseDomain
+	}
+
+	zone, err := client.GetDNSZone(cctx, ic.Config.GCP.ProjectID, domain, isPublic)
+	if err != nil {
+		return "", fmt.Errorf("failed to get dns zone name: %w", err)
+	}
+
+	if zone != nil {
+		return zone.Name, nil
+	}
+
+	return "", errNotFound
+}
+
+type recordSet struct {
+	projectID string
+	zoneName  string
+	record    *dns.ResourceRecordSet
+}
+
+// createRecordSets will create a list of records that will be created during the install.
+func createRecordSets(ctx context.Context, ic *installconfig.InstallConfig, clusterID, apiIP, apiIntIP string) ([]recordSet, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+
+	// A shared VPC install allows a user to preconfigure a private zone. If there is a private zone found,
+	// use the existing private zone otherwise default to the installer private zone pattern below.
+	privateZoneName, err := getDNSZoneName(ctx, ic, false)
+	if err != nil {
+		if !errors.Is(err, errNotFound) {
+			return nil, fmt.Errorf("failed to find private zone: %w", err)
+		}
+		privateZoneName = fmt.Sprintf("%s-private-zone", clusterID)
+	}
+
+	records := []recordSet{
+		{
+			// api_internal
+			projectID: ic.Config.GCP.ProjectID,
+			zoneName:  privateZoneName,
+			record: &dns.ResourceRecordSet{
+				Name:    fmt.Sprintf("api-int.%s.", ic.Config.ClusterDomain()),
+				Type:    "A",
+				Ttl:     60,
+				Rrdatas: []string{apiIntIP},
+			},
+		},
+		{
+			// api_external_internal_zone
+			projectID: ic.Config.GCP.ProjectID,
+			zoneName:  privateZoneName,
+			record: &dns.ResourceRecordSet{
+				Name:    fmt.Sprintf("api.%s.", ic.Config.ClusterDomain()),
+				Type:    "A",
+				Ttl:     60,
+				Rrdatas: []string{apiIntIP},
+			},
+		},
+	}
+
+	if ic.Config.Publish == types.ExternalPublishingStrategy {
+		existingPublicZoneName, err := getDNSZoneName(ctx, ic, true)
+		if err != nil {
+			if errors.Is(err, errNotFound) {
+				return nil, fmt.Errorf("failed to find a public zone: %w", err)
```

Comment:
This seem like similar errors returned, not sure its necessary to check `errNotFound`